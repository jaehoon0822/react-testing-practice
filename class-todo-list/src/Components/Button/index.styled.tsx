import styled from '@emotion/styled';
import { StyledButtonProps } from './index.types';

export const StyledButton = styled.div<StyledButtonProps>`
  display: flex;
  width: 100px;
  justify-content: center;
  align-items: center;
  outline: none;
  border: none;
  padding: 0.5rem;
  color: #fff;
  cursor: pointer;
  background-color: ${(props) => props.backgroundColor};
  transition: 0.2s all;
  &:hover {
    background-color: ${(props) => props.hoverColor};
  }
`;
