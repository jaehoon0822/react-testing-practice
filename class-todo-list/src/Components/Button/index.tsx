import { Component } from 'react';
import { Props } from './index.types';
import { StyledButton } from './index.styled';

export class Button extends Component<Props> {
  render() {
    const { label, backgroundColor = '#304ffe', hoverColor = '#1e40ff', onClick } = this.props;

    return (
      <StyledButton backgroundColor={backgroundColor} hoverColor={hoverColor} onClick={onClick} >
        {label}
      </StyledButton>
    );
  }
}
