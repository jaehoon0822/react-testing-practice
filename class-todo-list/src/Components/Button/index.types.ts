export interface Props {
  readonly label?: string;
  readonly backgroundColor?: string;
  readonly hoverColor?: string;
  readonly onClick?: () => void;
}

export interface StyledButtonProps extends Pick<Props, 'backgroundColor' | 'hoverColor'> {}
