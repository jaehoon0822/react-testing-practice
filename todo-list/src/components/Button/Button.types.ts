export interface ContainerProps {
  readonly backgroundColor: string;
  readonly hoverColor: string;
}

export interface Props {
  readonly label: string;
  readonly backgroundColor?: string;
  readonly hoverColor?: string;
  readonly onClick?: () => void;
}
