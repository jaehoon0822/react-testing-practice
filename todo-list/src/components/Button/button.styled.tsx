import styled from '@emotion/styled';
import { ContainerProps } from './button.types';

export const Container = styled.div<ContainerProps>`
  text-align: center;
  background-color: ${(props) => props.backgroundColor};
  padding: 10px 20px;
  border-radius: 8px;
  transition: 0.2s;
  cursor: pointer;
  &:hover {
    background-color: ${(props) => props.hoverColor};
  }
`;

export const Label = styled.div`
  color: #fff;
  font-size: 16px;
`;
