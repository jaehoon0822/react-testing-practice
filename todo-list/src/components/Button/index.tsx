import { Container, Label } from './Button.styled';
import { Props } from './Button.types';

const Button = ({ label, backgroundColor = '#304ffe', hoverColor = '#1e40ff', onClick }: Props) => {
  return (
    <Container backgroundColor={backgroundColor} hoverColor={hoverColor} onClick={onClick} data-testid={'container'}>
      <Label>{label}</Label>
    </Container>
  );
};

export default Button;
