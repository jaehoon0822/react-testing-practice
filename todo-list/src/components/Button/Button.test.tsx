import { render, screen } from '@testing-library/react';
import { matchers } from '@emotion/jest';
import Button from '.';
import userEvent from '@testing-library/user-event';

expect.extend(matchers);

describe('<Button />', () => {
  test('renders component correctly', async () => {
    // test 라는 이름을 가진 button render
    const { container } = render(<Button label="test" />);

    // test button 있는지 확인
    const label = screen.getByText('test');
    expect(label).toBeInTheDocument();

    // container element 선택
    const buttonContainer = screen.getByTestId('container');
    expect(buttonContainer).toBeInTheDocument();

    // button 상태에 따른 background-color 확인

    // hover 이전
    expect(buttonContainer).toHaveStyleRule('background-color', '#304ffe');
    // hover 이후
    expect(buttonContainer).toHaveStyleRule('background-color', '#1e40ff', { target: ':hover' });

    expect(container).toMatchSnapshot();
  });

  test('changes backgroundColor and hoveColor props', () => {
    // 기본 배경색상
    const backgroundColor = '#ff1744';
    // hover 되었을시 배경색상
    const hoverColor = '#f01440';

    // button component render
    render(<Button label="test" backgroundColor={backgroundColor} hoverColor={hoverColor} />);

    // 버튼 컨테이너 가져오기
    const buttonContainer = screen.getByTestId('container');
    expect(buttonContainer).toBeInTheDocument();

    // 버튼 container backgroundColor 확인
    expect(buttonContainer).toHaveStyleRule('background-color', backgroundColor);

    // 버튼 hover 시 backgroundColor 확인
    expect(buttonContainer).toHaveStyleRule('background-color', hoverColor, { target: ':hover' });
  });

  test('clicks the button', async () => {
    const user = userEvent.setup();

    // mock fn 생성
    const handleClick = jest.fn();

    // button 에 onClick event 로 mock fn 설정
    render(<Button label="test" onClick={handleClick} />);

    // onClick event 가 일어날 Element
    const buttonContainer = screen.getByTestId('container');

    // --> toHaveBeenCalledTimes 는 함수가 몇번 호출되었는지 확인하는 expect method.
    //
    // mock fn 이 몇번 클릭되었는지 확인: 0번
    expect(handleClick).toHaveBeenCalledTimes(0);

    // user.click 이벤트를 사용하여 해당 Element 클릭
    await user.click(buttonContainer);

    // mock fn 이 몇번 클릭되었는지 확인: 1번
    expect(handleClick).toHaveBeenCalledTimes(1);
  });
});
