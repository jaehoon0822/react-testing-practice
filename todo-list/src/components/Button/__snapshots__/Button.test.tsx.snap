// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> renders component correctly 1`] = `
<div>
  <div
    class="css-1jzh7nd"
    data-testid="container"
  >
    <div
      class="css-68lwrs"
    >
      test
    </div>
  </div>
</div>
`;
